File: sdk-python/dex/worker_options.py
# Copyright (c) 2026 Super Durable, Inc.
#
# Licensed under the Super Durable Source License 1.0.
# You may not use this file except in compliance with the License.
# See the LICENSE file in the repository root.
#
# SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

from dataclasses import dataclass
from datetime import timedelta


@dataclass(frozen=True)
class WorkerTarget:
    """Identify the application Worker endpoint advertised to Dex.

    Attributes:
        address: A plaintext gRPC target. Headless targets must use ``host:port``.
        headless: Whether Dex should connect directly without service discovery.
            Defaults to ``False``.
    """

    address: str
    headless: bool = False


@dataclass(frozen=True)
class WorkerOptions:
    """Configure Worker networking and startup index synchronization.

    ``attribute_index_sync_timeout`` is the RPC deadline used before the Worker
    opens its listener. It defaults to two minutes and must be positive.

    Attributes:
        bind_address: The local plaintext WorkerService listener in ``host:port``
            form. Defaults to ``":8803"``.
        worker_target: The endpoint advertised to Dex. ``None`` derives an address
            from ``bind_address``.
        server_address: The Dex FlowService target used for startup synchronization
            and blob hydration. Defaults to ``"localhost:8801"``.
        attribute_index_sync_timeout: Maximum duration for startup Attribute index
            synchronization. Defaults to two minutes and must be positive.
    """

    bind_address: str = ":8803"
    worker_target: WorkerTarget | None = None
    server_address: str = "localhost:8801"
    attribute_index_sync_timeout: timedelta = timedelta(minutes=2)

    def __post_init__(self) -> None:
        if self.attribute_index_sync_timeout <= timedelta(0):
            raise ValueError("attribute_index_sync_timeout must be positive")

File: sdk-rust/crates/dex-sdk/src/worker_options.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use std::time::Duration;

#[derive(Clone, Debug)]
pub struct WorkerTarget {
    address: String,
    headless: bool,
}

impl WorkerTarget {
    pub fn new(address: impl Into<String>) -> Self {
        Self {
            address: address.into(),
            headless: false,
        }
    }

    pub fn headless(mut self, value: bool) -> Self {
        self.headless = value;
        self
    }

    pub fn address(&self) -> &str {
        &self.address
    }

    pub fn is_headless(&self) -> bool {
        self.headless
    }
}

#[derive(Clone, Debug)]
pub struct WorkerOptions {
    bind_address: String,
    server_address: String,
    worker_target: Option<WorkerTarget>,
    attribute_index_sync_timeout: Duration,
}

impl WorkerOptions {
    pub fn new() -> Self {
        Self {
            bind_address: "0.0.0.0:8803".to_string(),
            server_address: "127.0.0.1:8801".to_string(),
            worker_target: None,
            attribute_index_sync_timeout: Duration::from_secs(120),
        }
    }

    pub fn bind_address(mut self, value: impl Into<String>) -> Self {
        self.bind_address = value.into();
        self
    }

    pub fn server_address(mut self, value: impl Into<String>) -> Self {
        self.server_address = value.into();
        self
    }

    pub fn worker_target(mut self, value: WorkerTarget) -> Self {
        self.worker_target = Some(value);
        self
    }

    /// Sets the startup Indexed Attribute synchronization deadline.
    ///
    /// The default is two minutes. [`Worker::try_new`](crate::Worker::try_new)
    /// rejects [`Duration::ZERO`].
    pub fn attribute_index_sync_timeout(mut self, value: Duration) -> Self {
        self.attribute_index_sync_timeout = value;
        self
    }

    pub(crate) fn bind_address_value(&self) -> &str {
        &self.bind_address
    }

    pub(crate) fn server_address_value(&self) -> &str {
        &self.server_address
    }

    pub(crate) fn worker_target_value(&self) -> Option<&WorkerTarget> {
        self.worker_target.as_ref()
    }

    pub(crate) fn attribute_index_sync_timeout_value(&self) -> Duration {
        self.attribute_index_sync_timeout
    }
}

impl Default for WorkerOptions {
    fn default() -> Self {
        Self::new()
    }
}
